### DISCUSSION TITLE
Using ThreadLocal in DGS

### DISCUSSION THREAD

**Discussion by pkadam1989** [Netflix/dgs-framework#720] [Q&A]
I want to add incoming graphql request headers to Threadlocal. 

Inside a datafetcher want to construct a RestTemplate client to reach out to another service with these headers. I want to register an interceptor for the RestTemplate which will read headers from ThreadLocal and add it to outgoing request. 

Since DGS datafetcher executes on a different thread, I won't have access to the Threadlocal object set in filter.

Need suggestion on how we can achieve this? Any suggestions will be helpful

**Comment by berngp** [comment]
I would avoid using `ThreadLocal` to store information. You could [use the DgsRequestData to get access to the request object](https://netflix.github.io/dgs/datafetching/#using-dgsrequestdata-to-get-access-to-the-request-object) and extract the headers as needed.

**Reply by pkadam1989** [reply]
Thank you for the quick response. 
I was able to extract the headers using DgsRequestData.getHeaders() method in DataFetcher.
`
DgsRequestData requestData =  DgsContext.getRequestData(dataFetchingEnvironment)
HttpHeaders headers = requestData.getHeaders();
`

I want to do the same thing in rest template interceptor https://www.baeldung.com/spring-rest-template-interceptor. 

How do I get access to the DataFetchingEnvironment object in my custom rest template interceptor? 
So that interceptor could copy these headers for the new outbound request

**Reply by berngp** [reply]
You would inject the RestTemplate to the DgsComponent/Service implementing the DataFetcher and call it in the scope of the method that has access to the context/request data.

**Reply by pkadam1989** [reply]
I have implemented with same approach. But did not want to attach headers to each request in every DataFetcher method.
So was trying to see if I can copy request headers from the DGS context/request data in an interceptor on the RestTemplate which will be common for all outbound requests.
